Azure Monitor: Filter namespaces by resource group

alyssajoyner opened the pull request:
Currently Azure Monitor fetches namespaces at a subscription level, making it difficult to know which namespaces are relevant to a given resource group when specified in a variable query. 

This PR adjusts the namespaces variable query to use resource graph instead of the Azure Monitor REST API. This allows the namespaces query to function for users with restricted permissions and will enable us to filter namespaces when a resource group is supplied.

alyssajoyner committed:
Update namespace endpoint to filter out only relevant namespaces

alyssajoyner committed:
Update tests

alyssajoyner committed:
Fix url builder tests

alyssajoyner committed:
Add todo comments

alyssajoyner committed:
Update func to use ARG to retrieve namespaces with metrics

alyssajoyner committed:
Refactor getMetricNamespaces for readability

alyssajoyner committed:
Lint

alyssajoyner committed:
Remove comments

alyssajoyner committed:
Remove type assertion

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/url_builder.ts:
```diff
@@ -68,9 +68,16 @@ export default class UrlBuilder {
+        region ? `&region=${region}` : globalRegion ? '&region=global' : ''
+      }`;
+    } else {
+      apiVersion = '2017-12-01-preview';
```
we didnt set this before. is this needed now?

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/url_builder.ts:
```diff
@@ -68,9 +68,16 @@ export default class UrlBuilder {
+      return `${baseUrl}${resourceUri}/resources?api-version=${apiVersion}${
+        region ? `&region=${region}` : globalRegion ? '&region=global' : ''
+      }`;
+    } else {
```
super nit: remove the else because we are returning in the if

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/url_builder.ts:
```diff
@@ -50,8 +50,8 @@ export default class UrlBuilder {
     apiVersion: string,
     query: GetMetricNamespacesQuery,
-    globalRegion: boolean,
     templateSrv: TemplateSrv,
+    globalRegion?: boolean,
```
nit: this function doesnt need `globalRegion` to be nullable. we shouldn't change this

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/response_parser.ts:
```diff
@@ -29,7 +29,7 @@ export default class ResponseParser {
       if (!find(list, ['value', get(result.value[i], valueFieldName)])) {
         const value = get(result.value[i], valueFieldName);
-        const text = get(result.value[i], textFieldName, value);
+        const text = get(result.value[i], textFieldName, value)?.toLowerCase();
```
why does this need to be lowercase now?

aangelisc replied:
I assume this needs to be lower case as the resource graph query may return them in varying cases - @alyssabull can you confirm?

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -238,47 +246,165 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<
+      globalRegion,
       region
     );
+
```
```suggestion
```

bossinc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -238,47 +246,165 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<
         if (custom) {
           result.value = result.value.filter((namespace) => namespace.classification === 'Custom');
         }
+
```
```suggestion
```

alyssajoyner committed:
Refactor ARG query

alyssajoyner committed:
Update tests and refactor class to use ARG

aangelisc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -238,47 +244,117 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<
           'properties.metricNamespaceName'
         );
       })
-      .then((result) => {
```
I don't think this can be removed, it will break how namespaces work for the sub-resources of Azure storage accounts unless something else has changed?

aangelisc commented on public/app/plugins/datasource/azuremonitor/azure_monitor/azure_monitor_datasource.ts:
```diff
@@ -238,47 +244,117 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<
+    }
+
+    // Default behavior if no resource group is provided
     const url = UrlBuilder.buildAzureMonitorGetMetricNamespacesUrl(
```
Rather than duplicating this code above I think we can use it once

aangelisc committed:
Substantial refactor

- Update Azure Resource Graph data source with a method for making paged requests and a method for retrieving metric namespaces (and add tests)
- Extract helpers from azure_monitor_datasource to utils and generalise them (also revert previous changes to datasource and test file)
- Update mock with Azure Resource Graph data source
- Revert response parser changes
- Revert url builder changes
- Update get metric namespaces query to use the resource graph method for variable queries
- Update docs

aangelisc committed:
Merge branch 'main' into alyssa/azure-monitor-filter-namespaces

aangelisc committed:
Lint

aangelisc committed:
Oops

aangelisc committed:
Fix type

aangelisc committed:
Lint and betterer

bossinc reviewed (approved):
Nice! It works great

beejeebus reviewed (approved):
LGTM.

lwandz13 commented on docs/sources/datasources/azure-monitor/template-variables/index.md:
```diff
@@ -48,18 +48,18 @@ For an introduction to templating and template variables, refer to the [Templati
+| ----------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| **Subscriptions**       | Returns subscriptions.                                                                                                                                                              |
+| **Resource Groups**     | Returns resource groups for a specified subscription. Supports multi-value.                                                                                                         |
+| **Namespaces**          | Returns metric namespaces for the specified subscription. If a resource group is supplied the namespaces returned will be filtered to those that are present in the resource group. |
```
```suggestion
| **Namespaces**          | Retrieves the metric namespaces for the specified subscription. If a resource group is provided, only the namespaces within that group are returned. |
```

lwandz13 reviewed (dismissed):
Looks great Alyssa! One small suggestion.

aangelisc committed:
Update docs/sources/datasources/azure-monitor/template-variables/index.md

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

aangelisc committed:
Merge branch 'main' into alyssa/azure-monitor-filter-namespaces

aangelisc committed:
Prettier write
